Aardvark/ARKEmailBugReporter.m

```diff
@@ -342,155 +336,128 @@ - (NSString *)formattedLogMessagesAttachmentExtension;
-    while (viewControllerToPresentAlertController.presentedViewController != nil) {
-        viewControllerToPresentAlertController = viewControllerToPresentAlertController.presentedViewController;
-    }
-
-    /*
-     Disabling animations here to avoid potential crashes resulting from unexpected view state in UIKit
-     */
-    [viewControllerToPresentAlertController presentViewController:alertController animated:NO completion:NULL];
-
 }
 
-- (void)_configureAlertTextfield:(UITextField *)textField
+- (ARKEmailBugReportConfiguration *)_configurationWithCurrentSettings;
 {
-    textField.autocapitalizationType = UITextAutocapitalizationTypeSentences;
-    textField.autocorrectionType = UITextAutocorrectionTypeYes;
-    textField.spellCheckingType = UITextSpellCheckingTypeYes;
-    textField.returnKeyType = UIReturnKeyDone;
-}
-
-- (void)_createBugReportWithTitle:(NSString *)title;
-{
-    NSArray *logStores;
+    ARKEmailBugReportConfiguration *const configuration = [[ARKEmailBugReportConfiguration alloc] init];
+    
     if (self.emailAttachmentAdditionsDelegate != nil) {
         NSMutableArray *const filteredLogStores = [NSMutableArray arrayWithCapacity:self.logStores.count];
         for (ARKLogStore *logStore in self.logStores) {
             if ([self.emailAttachmentAdditionsDelegate emailBugReporter:self shouldIncludeLogStoreInBugReport:logStore]) {
                 [filteredLogStores addObject:logStore];
             }
         }
-        logStores = filteredLogStores;
+        configuration.logStores = filteredLogStores;
+        
+        configuration.additionalAttachments = [self.emailAttachmentAdditionsDelegate additionalEmailAttachmentsForEmailBugReporter:self];
+        
     } else {
-        logStores = [self.logStores copy];
+        configuration.logStores = [self.logStores copy];
     }
     
+    configuration.includesScreenshot = self.attachScreenshotToNextBugReport;
+    configuration.includesViewHierarchyDescription = (self.attachScreenshotToNextBugReport && self.attachesViewHierarchyDescriptionWithScreenshot);
+    
+    return configuration;
+}
+
+- (void)_createBugReportWithConfiguration:(ARKEmailBugReportConfiguration *)configuration;
+{
     NSMapTable *logStoresToLogMessagesMap = [NSMapTable new];
     NSDictionary *emailBodyAdditions = [self.emailBodyAdditionsDelegate emailBodyAdditionsForEmailBugReporter:self];
     
     if ([MFMailComposeViewController canSendMail]) {
         self.mailComposeViewController = [MFMailComposeViewController new];
         
         [self.mailComposeViewController setToRecipients:@[self.bugReportRecipientEmailAddress]];
-        [self.mailComposeViewController setSubject:title];
+        [self.mailComposeViewController setSubject:configuration.prefilledEmailSubject];
         
+        dispatch_group_t logStoreRetrievalDispatchGroup = dispatch_group_create();
+        dispatch_group_enter(logStoreRetrievalDispatchGroup);
```

Comment:
nit: technically there's no need for this outside `enter`/`leave`. As long as you're calling `notify` after you've `enter`ed all your async tasks, you're fine.

that said, I'm pretty sure the surrounding `enter`/`leave` is in the style of other code here, so, we're good either way.

Reply:
I think theoretically it's better because we could change `retrieveAllLogMessagesWithCompletionHandler` to be synchronous at some point and potentially inadvertently change behavior here.

Reply:
As long as `notify` is called after current async work, and after you'd want it to be executed, any changes to sync/async behavior above shouldn't cause a change.

But yeah, the wrapper certainly makes things a bit more foolproof :)